src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -728,162 +760,215 @@ DatabaseShardImp::onChildrenStopped()
 void
 DatabaseShardImp::import(Database& source)
 {
+    assert(init_);
+
+    // Only the application local node store can be imported
+    if (&source != &app_.getNodeStore())
     {
-        std::lock_guard lock(mutex_);
-        assert(init_);
+        assert(false);
+        JLOG(j_.error()) << "invalid source database";
+        return;
+    }
+
+    std::uint32_t earliestIndex;
+    std::uint32_t latestIndex;
+    {
+        auto loadLedger = [&](bool ascendSort =
+                                  true) -> boost::optional<std::uint32_t> {
+            std::shared_ptr<Ledger> ledger;
+            std::uint32_t ledgerSeq;
+            std::tie(ledger, ledgerSeq, std::ignore) = loadLedgerHelper(
+                "WHERE LedgerSeq >= " + std::to_string(earliestLedgerSeq()) +
+                    " order by LedgerSeq " + (ascendSort ? "asc" : "desc") +
+                    " limit 1",
+                app_,
+                false);
+            if (!ledger || ledgerSeq == 0)
+            {
+                JLOG(j_.error()) << "no suitable ledgers were found in"
+                                    " the SQLite database to import";
+                return boost::none;
+            }
+            return ledgerSeq;
+        };
+
+        // Find earliest ledger sequence stored
+        auto ledgerSeq{loadLedger()};
+        if (!ledgerSeq)
+            return;
+        earliestIndex = seqToShardIndex(*ledgerSeq);
 
-        // Only the application local node store can be imported
-        if (&source != &app_.getNodeStore())
+        // Consider only complete shards
+        if (ledgerSeq != firstLedgerSeq(earliestIndex))
+            ++earliestIndex;
+
+        // Find last ledger sequence stored
+        ledgerSeq = loadLedger(false);
+        if (!ledgerSeq)
+            return;
+        latestIndex = seqToShardIndex(*ledgerSeq);
+
+        // Consider only complete shards
+        if (ledgerSeq != lastLedgerSeq(latestIndex))
+            --latestIndex;
+
+        if (latestIndex < earliestIndex)
         {
-            assert(false);
-            JLOG(j_.error()) << "invalid source database";
+            JLOG(j_.error()) << "no suitable ledgers were found in"
+                                " the SQLite database to import";
             return;
         }
+    }
+
+    auto numHistShards = numHistoricalShards(std::lock_guard(mutex_));
+    auto const sweepInterval = std::chrono::milliseconds(500);
+
+    // Sweep the source and family caches to conserve
+    // memory
+    ThreadWrapper mainSweeper(
+        [&source, sweepInterval, &family = app_.getNodeFamily()] {
+            source.sweep();
```

Comment:
The node store sweep does nothing, no need to call it.

Reply:
I needed this before Joel's database modifications. I will remove the sweeps and see how the server performs.

Reply:
The family sweep is still needed. Done

Reply:
What is the memory consumption without the family sweep?

Reply:
> What is the memory consumption without the family sweep?

I saw it get up to around 13GB. I stepped away and by the time I returned the process had been killed.